### PR TITLE
44708627_Efficient_Sub_Pixel_CNN

### PR DESCRIPTION
**Pull request by LingxiaoGao**
The model implemented is able to reconstructs a high-resolution version of an image from a low-resolution (factor of 4) version given.

The working principles were to train a customized CNN network with tuple of (low-res image, high-res image) and optimize mean squared error between prediction and high-res image.

Peak signal-to-noise ratio (PSNR) is commonly used to quantify reconstruction quality for images and is defined via MSE. Lower MSE indicates higher PSNR and reconstruction quality.

The basic structure was inspired by ESPCN (Efficient Sub-Pixel CNN) proposed by Shi, 2016 and Keras implementation by Long, 2020. The model was also embedded with Residual Dense Blocks, inspired by Chakraborty, 2021. Residual blocks prevented early saturation or degradation with accuracy from increasing number of layers.

### CODE REVIEW

**Comment by uqmarlonbran** [comment]
Performance seems in line with most other implementations. All seems ok.

**Comment by LingxiaoGao** [comment]
> Performance seems in line with most other implementations. All seems ok.

This was built upon keras implementation, which I believe most people would also do this as this is provided in Task sheet.
Keras implementation has already achieved a high PSNR, and adding convolution layers straightforward or modifying hyper parameters just make the performance worse. 

I have modified the base model of the network and increases the PSNR by 2-3 compared with original implementation, using a customized residual units before the output layer.

**Comment by shakes76** [comment]
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate use and implementation (convert from notebook, typo in module file name) -2
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves problem
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage (missing plots) -1
Module present
Commenting
No Data leakage
Difficulty: Normal -5
## Commit Log
Meaningful commit messages -1 (could be more informative)
Progressive commits used
## Documentation
ReadMe acceptable/good
Good Description and Comments 
Markdown used PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
Feedback required, remove model weights -2
Request Description good
